blog: add Start sidecar first (#50673)
kubernetes/website: content/en/blog/_posts/2025-05-09-start-sidecar-first/index.md

---
layout: blog
title: "Start Sidecar First: How To Avoid Snags"
date: 2025-05-05
draft: true
slug: start-sidecar-first
author: Agata Skorupka (The Scale Factory)
---

> From the [previous blog post](/blog/2025/04/22/multi-container-pods-overview/) on Multicontainer Pods you know what their job is, what are the main architectural patterns, and how they are implemented in Kubernetes. The main thing I’ll cover in this article is how to ensure that your sidecar containers start before the main app. It’s more complicated than you might think!

Author: `previous blog post` replaced, but I've left the "thing" as I like it more.

## A gentle refresher

> Let’s just remind that the [v1.29.0 release of Kubernetes](/blog/2023/12/13/kubernetes-v1-29-release/) added native support for [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/), which can now be defined within the `spec.initContainers` field, but with `restartPolicy: Always` as in the example below (note that it is a partial snippet of the full Kubernetes manifest):

Reviewer: s/Let’s just remind that/ As a reminder, /

Reviewer: > as in the example below (note that it is a partial snippet of the full Kubernetes manifest)

Consider replacing with something like "the following example Pod manifest snippet...".

```yaml
initContainers:
  - name: logshipper
    image: alpine:latest
    restartPolicy: Always
    command: ['sh', '-c', 'tail -F /opt/logs.txt']
    volumeMounts:
    - name: data
        mountPath: /opt
```

> What’s the specifics of defining sidecars with a `spec.initContainers` block, rather than as a regular multi-container pod? `spec.initContainers` are always started before the main application and terminated after the main application. Furthermore, when used with [Jobs](/docs/concepts/workloads/controllers/job/), a sidecar container is still alive and potentially restarts after the Job is done, and they will not block pod completion (see also [Pod sidecar containers tutorial](/docs/tutorials/configuration/pod-sidecar-containers/)).

Reviewer 2: Optional, recommended wording tweak:
```suggestion
What are the specifics of defining sidecars with a `.spec.initContainers` block, rather than as a legacy multi-container pod with multiple `.spec.containers`?
Well, all `.spec.initContainers` are always launched **before** the main application. If you define Kubernetes-native sidecars, those are terminated **after** the main application. Furthermore, when used with [Jobs](/docs/concepts/workloads/controllers/job/), a sidecar container should still be alive and could potentially even restart after the owning Job is complete; Kubernetes-native sidecar containers do not block pod completion.

To learn more, you can also read the official [Pod sidecar containers tutorial](/docs/tutorials/configuration/pod-sidecar-containers/).
```

> What’s the specifics of defining sidecars with an `spec.initContainers` block, rather than as a regular multi-container pod? `spec.initContainers` are always started first (before the main application) and terminated last (after the main application). Furthermore, when used with [Jobs](/docs/concepts/workloads/controllers/job/), sidecar would still be alive and potentially restarted after Job is done, and they will not block pod completion (see also [Pod sidecar containers tutorial](/docs/tutorials/configuration/pod-sidecar-containers/)).

Reviewer: s/What’s the/What are the/

s/an `spec.initContainers`/a `spec.initContainers`/

> are always started first (before the main application) and terminated last (after the main application).

Consider simplifying and being more specific, For example:

`are always started before the main application and terminated after the main application.`

s/sidecar would still be alive and potentially restarted after Job.../a sidecar container is still alive and potentially restarts after the Job.../

## The problem

> Now you know that defining a sidecar with this native approach will always start it before the main application. From the kubelet source code, we know that this often means [being started almost in parallel](https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74), and this is not always what we want to achieve. What we are really interested in is whether we can delay the start of the main application until the sidecar is not just started, but fully running and ready to serve.

Reviewer: > Now you know that defining a sidecar with this native approach will always start it before the main application.

Consider rewording, e.g. "Now that you know a sidecar container always starts before the main..."

> From the kubelet source code, we know that this often means [being started almost in parallel](https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74)

Why not reference the source code instead of this 3rd party blog post?

Author: I agree referencing source code looks better so I've changed that, but to be honest I wanted to give the author of that article a reference.

> Now you know that defining a sidecar with this native approach will always start it before the main application. From the [kubelet source code](https://github.com/kubernetes/kubernetes/blob/537a602195efdc04cdf2cb0368792afad082d9fd/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L827-L830), we know that this often means being started almost in parallel, and this is not always what we want to achieve. What we are really interested in is whether we can delay the start of the main application until the sidecar is not just started, but fully running and ready to serve.

Reviewer 2: Avoid "we" here.

It might be a bit tricky because the problem with sidecars is there’s no obvious success signal, contrary to init containers - designed to run only for a specified period of time. With an init container, exit status 0 is unambiguously "I succeeded". With a sidecar, there are lots of points at which you can say "a thing is running".

> Starting one container only after the previous one is ready is part of a graceful deployment strategy, ensuring proper sequencing and stability during startup. It’s also actually how I’d expect sidecar containers to work as well, as it would cover the scenario where the main application is dependent on the sidecar. For example, it may happen that an app errors out if the sidecar isn’t available to serve requests (e.g., logging with DataDog). Sure, we could change the application code (and it would actually be the “best practice” solution), but sometimes we can’t - and this post focuses on this use case.

Reviewer: > It’s also actually how I’d expect sidecar containers to work as well, as it would cover the scenario where the main application is dependent on the sidecar.

Consider rewording for better clarity.

> Starting one container only after the previous one is ready is part of a graceful deployment strategy, ensuring proper sequencing and stability during startup. It’s also actually how I’d expect sidecar containers to work as well, to cover the scenario where the main application is dependent on the sidecar. For example, it may happen that an app errors out if the sidecar isn’t available to serve requests (e.g., logging with DataDog). Sure, we could change the application code (and it would actually be the “best practice” solution), but sometimes we can’t - and this post focuses on this use case.

Reviewer 2: Avoid "we" here.

## Readiness probe

> To check whether Kubernetes native sidecar delays the start of the main application until sidecar is ready, let’s simulate a short investigation. Firstly, we’ll simulate a sidecar container which will never be ready by implementing a readiness probe which will never succeed. As a reminder, a readiness probe checks if the container is ready to start accepting traffic and therefore, if the pod can be used as a backend for services. 

Reviewer 3: ```suggestion
To check whether Kubernetes native sidecar delays the start of the main application until sidecar is ready, let’s simulate a short investigation. Firstly, we’ll simulate a sidecar container which will never be ready by implementing a readiness probe which will never succeed. As a reminder, a [readiness probe](/docs/concepts/configuration/liveness-readiness-startup-probes/) checks if the container is ready to start accepting traffic and therefore, if the pod can be used as a backend for services. 
```

> To check whether Kubernetes native sidecar delays the start of the main application until sidecar is ready, let’s simulate a short investigation. Firstly, we’ll simulate a sidecar container which will never be ready by implementing a readiness probe which will never succeed. As a reminder, a [readiness probe](/docs/concepts/configuration/liveness-readiness-startup-probes/) checks if the container is ready to start accepting traffic and therefore, if the pod can be used as a backend for services. 

Reviewer: s/sidecar/the sidecar/

> (Unlike true init containers, sidecar containers can have [probes](https://kubernetes.io/docs/concepts/configuration/liveness-readiness-startup-probes/) so that the kubelet can supervise the sidecar and intervene if there are problems. For example, restarting a sidecar container if it fails a health check.)

Reviewer: s/true/standard/

(...)
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  17s               default-scheduler  Successfully assigned default/myapp-db5474f45-htgw5 to node01
  Normal   Pulling    16s               kubelet            Pulling image "nginx:latest"
  Normal   Pulled     16s               kubelet            Successfully pulled image "nginx:latest" in 163ms (163ms including waiting). Image size: 72080558 bytes.
  Normal   Created    16s               kubelet            Created container nginx
  Normal   Started    16s               kubelet            Started container nginx
  Normal   Pulling    15s               kubelet            Pulling image "alpine:latest"
  Normal   Pulled     15s               kubelet            Successfully pulled image "alpine:latest" in 159ms (160ms including waiting). Image size: 3652536 bytes.
  Normal   Created    15s               kubelet            Created container myapp
  Normal   Started    15s               kubelet            Started container myapp
  Warning  Unhealthy  1s (x6 over 15s)  kubelet            Readiness probe failed:
```

> From the above findings it’s evident that only one container is ready - and I know it can’t be the sidecar, because I’ve defined it so it’ll never be ready. I also saw that myapp has been started before the sidecar is ready. That was not the result I wanted to achieve; in this case, the main app container has a hard dependency on its sidecar.

Reviewer: > From the above findings it’s evident that only one container is ready - and I know it can’t be the sidecar, because I’ve defined it so it’ll never be ready.

Consider rewording for clarity. Also, are there any other commands, events, etc., available that indicate which container is failing the probe?

## Maybe a startup probe?

> To ensure that the sidecar is ready before the main app container starts, I can define a `startupProbe`. It will delay the start of the main container till the successful execution of the command. If you’re wondering why I’ve added it to my `initContainer`, let’s analyse what would happen If I’d added it to myapp container. I wouldn’t have guaranteed the probe would run before the main application code - and this one, can potentially error out without the sidecar being up and running.

Reviewer: s/till the successful execution of the command/until the command is successfully executed/

What is "successfully executed"? The command returns a `0` exit code, correct?

Reviewer: s/would happen/happens/

Present tense should be used according to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

Author: yes, added that clarification in the text

```yaml                                                                
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myapp
  labels:
    app: myapp
spec:
  replicas: 1
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
    spec:
      containers:
        - name: myapp
          image: alpine:latest
          command: ["sh", "-c", "sleep 3600"]
      initContainers:
        - name: nginx
          image: nginx:latest
          ports:
            - containerPort: 80
              protocol: TCP
          restartPolicy: Always
          startupProbe:
            httpGet:
              path: /
              port: 80
            initialDelaySeconds: 5
            periodSeconds: 30
            failureThreshold: 10
            timeoutSeconds: 20
      volumes:
        - name: data
          emptyDir: {}
```

> This resulted in 2/2 containers being ready and running, and from events, it can be inferred that the main application started only after nginx had already been started. But to confirm whether it waited for the sidecar readiness, let’s change the `startupProbe` to the exec type of command: 

Reviewer: s/resulted/results/

https://kubernetes.io/docs/contribute/style/style-guide/

```yaml
startupProbe:
  exec:
    command:
    - /bin/sh
    - -c
    - sleep 15
```

and run `kubectl get pods -w` to watch in real time whether the readiness of both containers only changes after a 15 second delay. Again, events confirm the main application starts after the sidecar.

> That means that using the `startupProbe` with a correct httpGet request helps to delay the main application start until the sidecar is ready. It’s not optimal, but it works (I can hope this functionality will be added to Kubernetes-native sidecar at some point).

Reviewer 3: ```suggestion
That means that using the `startupProbe` with a correct `startupProbe.httpGet` request helps to delay the main application start until the sidecar is ready. It’s not optimal, but it works.
```

## What about the postStart lifecycle hook?

Fun fact: using the `postStart` lifecycle hook block will also do the job, but I’d have to write my own mini-shell script, which is even less efficient.

```yaml
initContainers:
  - name: nginx
    image: nginx:latest
    restartPolicy: Always
    ports:
      - containerPort: 80
        protocol: TCP
    lifecycle:
      postStart:
        exec:
          command:
          - /bin/sh
          - -c
          - |
            echo "Waiting for readiness at http://localhost:80"
            until curl -sf http://localhost:80; do
              echo "Still waiting for http://localhost:80..."
              sleep 5
            done
            echo "Service is ready at http://localhost:80"
```

## Liveness probe

> An interesting exercise would be to check the sidecar container behavior with the liveness probe. The liveness probe works and is configured similarly to the readiness probe - just with the difference that it doesn’t affect the readiness of the container but restarts it in case the probe fails. 

Reviewer 3: ```suggestion
An interesting exercise would be to check the sidecar container behavior with the [liveness probe](/docs/concepts/configuration/liveness-readiness-startup-probes/). The liveness probe works and is configured similarly to the readiness probe - just with the difference that it doesn’t affect the readiness of the container but restarts it in case the probe fails. 
```

```yaml
livenessProbe:
  exec:
    command:
    - /bin/sh
    - -c
    - exit 1 # this command always fails, keeping the container "Not Ready"
  periodSeconds: 5
```

> After adding the liveness probe configured just as the previous readiness probe and checking events of the pod by `kubectl describe pod` it’s visible that the sidecar has a restart count above 0. Nevertheless, the main application is not restarted nor influenced at all, even though we’re aware that (in our imaginary worst-case scenario) it can error out when the sidecar is not there serving requests.

Reviewer 2: Avoid "we" here.

## Findings summary

I’ll summarize the startup behavior in the table below:

| Probe/Hook     | Sidecar starts before the main app?                      | Main app waits for the sidecar to be ready?         | What if the check doesn’t pass?                    |
|----------------|----------------------------------------------------------|-----------------------------------------------------|----------------------------------------------------|
| readinessProbe | Yes, but it’s almost in parallel (effectively no)        | No                                                  | Sidecar is not ready, main app is running          |
| livenessProbe  | Yes, but it’s almost in parallel (effectively no)        | No                                                  | Sidecar is restarted, main app is running          |
| startupProbe   | Yes, but it’s almost in parallel (effectively no)        | Yes                                                 | Main app is not started                            |

> | postStart      | Yes, sequentially after postStart is executed            | Yes, but you have to provide custom logic for that  | Main app is not started                            |

Reviewer 2: (optional tweak)
To help readers, you could write the words Yes and No both in boldface.

> To wrap up, with sidecars being often dependencies for the main application, you may want to delay the start of the latter till the sidecar is ready. The ideal pattern is to start both containers simultaneously and have the app container logic delay at all levels, but it’s not always possible. In such a case, we have to introduce a bit of custom logic to our resource definition. Thankfully, it’s nice and quick, and you have the recipe ready above. 

Reviewer: s/sidecars being often dependencies for/sidecars often being a dependency of/

Reviewer: s/till/until/

What if I’d used a `livenessProbe` without lifecycle `postStart`? Both containers will be immediately ready: at the beginning, this behavior will not be different from the one without any additional probes since the liveness probe doesn’t affect readiness at all. After a while, the sidecar will begin to restart itself, but it won’t influence the main container.

## Findings summary

I’ll summarize the startup behavior in the table below:

| Probe/Hook     | Sidecar starts before the main app?                      | Main app waits for the sidecar to be ready?         | What if the check doesn’t pass?                    |
|----------------|----------------------------------------------------------|-----------------------------------------------------|----------------------------------------------------|

> | readinessProbe | Yes, but it’s almost in parallel (effectively **no**)        | **No**                                                  | Sidecar is not ready, main app is running          |
> | livenessProbe  | Yes, but it’s almost in parallel (effectively **no**)        | **No**                                                  | Sidecar is restarted, main app is running          |
> | startupProbe   | Yes, but it’s almost in parallel (effectively **no**)        | **Yes**                                                 | Main app is not started                            |
> | postStart      | **Yes**, sequentially after postStart is executed            | **Yes**, but you have to provide custom logic for that  | Main app is not started                            |

Reviewer 2: Is this change more accurate:
```suggestion
| `readinessProbe` | **Yes**, but it’s almost in parallel (effectively **no**)    | **No**                                                  | Sidecar is not ready; main app continues running   |
| `livenessProbe`  | Yes, but it’s almost in parallel (effectively **no**)    | **No**                                                  | Sidecar is restarted, main app continues running   |

| `startupProbe`   | **Yes**                                                      | **Yes**                                                 | Main app is not started                            |
| postStart      | **Yes**, main app container starts after `postStart` completes | **Yes**, but you have to provide custom logic for that  | Main app is not started
```

Reviewer 2: If it is, let's try to fix that before publication.
